protocol/daemons/server/server.go

```diff
@@ -22,32 +27,74 @@ type Server struct {
 	fileHandler   lib.FileHandler
 	socketAddress string
 
+	updateMonitor *types.UpdateMonitor
+
 	BridgeServer
 	PriceFeedServer
 	LiquidationServer
 }
 
 // NewServer creates a single gRPC server that's shared across multiple daemons for communication.
+// uniqueTestIdentifier is a string that can be used to identify services spawned by a particular test case,
+// so that they can be cleaned up after the test case is complete.
 func NewServer(
 	logger log.Logger,
 	grpcServer lib.GrpcServer,
 	fileHandler lib.FileHandler,
 	socketAddress string,
+	uniqueTestIdentifier string,
 ) *Server {
 	srv := &Server{
 		logger:        logger,
 		gsrv:          grpcServer,
 		fileHandler:   fileHandler,
 		socketAddress: socketAddress,
+		updateMonitor: types.NewUpdateFrequencyMonitor(),
 	}
+	stoppable.RegisterServiceForTestCleanup(uniqueTestIdentifier, srv)
 	return srv
 }
 
 // Stop stops the daemon server's gRPC service.
 func (server *Server) Stop() {
+	server.updateMonitor.Stop()
 	server.gsrv.Stop()
 }
 
+// registerDaemon registers a daemon service with the update monitor.
+func (server *Server) registerDaemon(
+	daemonKey string,
+	maximumAcceptableUpdateDelay time.Duration,
+) {
+	go func() {
+		// Give the protocol time to come up and start the gRPC query service. The daemons are unable
+		// to complete startup until this connection is established.
+		time.Sleep(DaemonStartupGracePeriod)
+		if err := server.updateMonitor.RegisterDaemonService(daemonKey, maximumAcceptableUpdateDelay); err != nil {
+			server.logger.Error("Failed to register daemon service with update monitor", "error", err)
```

Comment:
The registration can fail in test cases where the test completes in less time than the daemon startup period.

Reply:
Hmm can think deeper about a clean solution here, but somewhat uncomfortable allowing the daemon to not run here.. wouldn't this run into the same issue this PR is trying to prevent?

Understand this is an edge case so we can potentially create a ticket to fix this later. However ideally if the fix is simple, we can fix it now.

Reply:
Check out the fix!